GTBitsOfGood/fair-districts-ga
# Make newspaper page

## Construction of newspaper page

Issue Number(s): #12 

Created `newspaper` and associated `/api/newspaper` API route. Ultimately selected ChakraUI for fast prototyping of form, modal, and table elements as design was pending. 

### Additional Notes
- Adding counties through the form checks whether a county with the name already exists upon submission
    - Empty string elements are filtered out, County relations are created dynamically when user inputs new county name
    - Upon adding a county to a newspaper and the county already exists, the `County.Newspapers` attribute is appended with the edited newspaper
    - Adding/removing counties modifies all related fields (`Newspaper.Counties` and `County.Newspapers`)
- Deletion of rows (newspapers) occurs through the edit modal
- Database errors prevent the form from submitting and the modal from closing, possible error message 
    - Potential database errors in testing: adding a newspaper with the same name

### Checklist

- [x]  Adding a newspaper should be reflected in the database
- [x]  Adding a newspaper should display the information in a table
- [x]  The form should produce an error message if a required field is not inputted
- [x]  The table should be scrollable once enough users are added
- [x]  When going back to the page, table is persistent

### Critical Changes
- Global prisma instance to prevent creation of multiple prisma instances throughout lifecycle of app: `prisma/prisma.js`
- BREAKING CHANGE in b58077b8f61e6451025812132408270fe6a6ee2a that makes county names unique
    - Requires migration
- ChakraUI provider in `pages/_app.js`

### How to Test

Add, edit, and delete newspapers — some with the same counties — at will. Ensure that CRUD operations perform as expected (correct index updated/deleted, form validation).
